src/main/scala/ch/epfl/bluebrain/nexus/kg/async/ProjectDigestCoordinatorActor.scala

```diff
@@ -0,0 +1,206 @@
+import akka.cluster.sharding.{ClusterSharding, ClusterShardingSettings, ShardRegion}
+import akka.pattern.pipe
+import akka.persistence.query.{NoOffset, Offset, Sequence, TimeBasedUUID}
+import cats.MonadError
+import cats.implicits._
+import ch.epfl.bluebrain.nexus.admin.client.types.Project
+import ch.epfl.bluebrain.nexus.kg.KgError
+import ch.epfl.bluebrain.nexus.kg.KgError.InternalError
+import ch.epfl.bluebrain.nexus.kg.async.ProjectDigestCoordinatorActor.Msg._
+import ch.epfl.bluebrain.nexus.kg.async.ProjectDigestCoordinatorActor.OffsetSyntax
+import ch.epfl.bluebrain.nexus.kg.config.AppConfig
+import ch.epfl.bluebrain.nexus.kg.config.AppConfig._
+import ch.epfl.bluebrain.nexus.kg.indexing.Statistics
+import ch.epfl.bluebrain.nexus.kg.resources.Rejection.{FileDigestNotComputed, UnexpectedState}
+import ch.epfl.bluebrain.nexus.kg.resources.syntax._
+import ch.epfl.bluebrain.nexus.kg.resources.{Event, Files, Resource}
+import ch.epfl.bluebrain.nexus.kg.storage.Storage.StorageOperations.FetchDigest
+import ch.epfl.bluebrain.nexus.sourcing.projections.ProjectionProgress.NoProgress
+import ch.epfl.bluebrain.nexus.sourcing.projections._
+import kamon.Kamon
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import monix.execution.atomic.AtomicLong
+
+/**
+  * Coordinator backed by akka actor which runs the digest stream inside the provided project
+  */
+//noinspection ActorMutableStateInspection
+private abstract class ProjectDigestCoordinatorActor(implicit val config: AppConfig) extends Actor with ActorLogging {
+
+  private var child: Option[StreamSupervisor[Task, ProjectionProgress]] = None
+
+  def receive: Receive = {
+    case Start(_, project: Project) =>
+      log.debug("Started digest coordinator for project '{}'", project.projectLabel.show)
+      context.become(initialized(project))
+      child = Some(startCoordinator(project, restartOffset = false))
+    case FetchProgress(_) => val _ = progress.runToFuture pipeTo sender()
+    case other =>
+      log.debug("Received non Start message '{}', ignore", other)
+  }
+
+  def initialized(project: Project): Receive = {
+    case Stop(_) =>
+      log.info("Digest process for project '{}' received a stop message.", project.projectLabel.show)
+      child.foreach(_.stop())
+      child = None
+      context.become(receive)
+
+    case FetchProgress(_) => val _ = progress.runToFuture pipeTo sender()
+
+    case _ => //ignore
+  }
+
+  def startCoordinator(project: Project, restartOffset: Boolean): StreamSupervisor[Task, ProjectionProgress]
+
+  private def progress: Task[Statistics] =
+    child.map(_.state().map(_.getOrElse(NoProgress))).getOrElse(Task.pure(NoProgress)).map { p =>
+      Statistics(p.processedCount, p.discardedCount, p.processedCount, p.offset.asInstant, p.offset.asInstant)
+    }
+}
+
+private class FileDigestEventMapping[F[_]: FetchDigest](files: Files[F])(implicit F: MonadError[F, KgError]) {
```

Comment:
Because as you can see I only use this MonadError to throw errors of subtype `KgError`. And because the TagProjections will only retry for those...so the more concise the better, right?